### PR TITLE
Apply gulp-changed to style and script tasks

### PR DESCRIPTION
**Pull request by Scotchester**
Adds [gulp-changed](https://github.com/sindresorhus/gulp-changed) to style and script tasks to prevent them from being run if the files at the destination are older than the source files.

Discussion welcome!

## Testing

1. Pull branch
2. `npm install`
3. Run `gulp styles` and note how long each subtask takes. (Don't rely on the final overall `styles` time, as it doesn't seem to include the time taken on each subtask.)
4. Run `gulp styles` again and note how it takes almost no time, since none of the styles need to be rebuilt!
5. Run `gulp scripts` a couple times and watch the time go way down.
6. Test various pages (especially things that use some of the one-off style and script tasks) to ensure they are still working

## Todos

- [x] Investigate testing the destination file within each different Gulp sub-task.
- [x] Add gulp-changed to script tasks (and others, if appropriate)

## Checklist

* [ ] PR has an informative and human-readable title
* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [development playbook](https://github.com/cfpb/development)
* [ ] Passes all existing automated tests
* [ ] Any *change* in functionality is tested
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated
* [ ] Reviewers requested with the [Reviewer tool](https://help.github.com/articles/about-pull-request-reviews/) :arrow_right:

### CODE REVIEW

File: gulp/tasks/clean.js [hunk 1]
```diff
@@ -4,6 +4,25 @@ const configClean = require( '../config' ).clean;
 const del = require( 'del' );
 const gulp = require( 'gulp' );
 
+
+/**
+ * Cleans out built CSS files
+ */
+function cleanCSS() {
+  del( configClean.css + '/**/*' );
```
Comment (sebworks) [inline gulp/tasks/clean.js]: Add a return statement here

File: gulp/tasks/clean.js [hunk 2]
```diff
@@ -4,6 +4,25 @@ const configClean = require( '../config' ).clean;
...
+}
+
+/**
+ * Cleans out built JavaScript files
+ */
+function cleanJS() {
+  del( configClean.js + '/**/*' );
```
Comment (sebworks) [inline gulp/tasks/clean.js]: Same.

**Comment by Scotchester** [comment]
Pushed an update that adds [gulp-changed](https://github.com/sindresorhus/gulp-changed) to the project and uses it in each style subtask to ensure that it will only run if the destination file(s) are older than the source file(s). Testing steps updated above!

File: gulp/tasks/clean.js (hunk 1 above)
Comment (Scotchester) [reply gulp/tasks/clean.js]: What should it return?

Comment (sebworks) [reply gulp/tasks/clean.js]: sorry, so it should just return `del( configClean.css + '/**/*' );`

Comment (sebworks) [reply gulp/tasks/clean.js]: https://github.com/sindresorhus/del/issues/32

File: gulp/config.js [hunk 3]
```diff
@@ -42,6 +42,8 @@ module.exports = {
     reporter: process.env.CONTINUOUS_INTEGRATION // eslint-disable-line no-process-env
   },
   clean: {
+    css: paths.processed + '/css',
+    js: paths.processed + '/js',
```
Comment (jimmynotjim) [inline gulp/config.js]: Minor: Left align these values to match our existing style.

File: frontend.sh [hunk 4]
```diff
@@ -85,11 +85,32 @@ clean_and_install() {
 # Run tasks to build the project for distribution.
 build() {
   echo 'Building project…'
-  gulp clean
-  gulp build
+
+  if [ cfgov/static_built/css/main.css -ot cfgov/unprocessed/css/ ]; then
```
Comment (sebworks) [inline frontend.sh]: it would be nice if this could use though config file, I know it's a fantasy though.

**Comment by sebworks** [comment]
looks good to me, pending adding the return statement.

**Comment by Scotchester** [comment]
Working on adding it to the JS tasks, too, before merging. Figuring out how to do it with the Modernizr build is the main headscratcher.

File: cfgov/jinja2/v1/_layouts/base.html [hunk 5]
```diff
@@ -263,9 +263,9 @@
     {% compress js %}
     <script>
     {% for js in page.media %}
-       {% set compononet_js_source = component_ondemand_js(js) | trim %}
-       {% if compononet_js_source | length > 0 %}
-            {{ compononet_js_source }}
+       {% set compononent_js_source = component_ondemand_js(js) | trim %}
+       {% if compononent_js_source | length > 0 %}
+            {{ compononent_js_source }}
```
Comment (Scotchester) [inline cfgov/jinja2/v1/_layouts/base.html]: The changes in this file are just fixing a typo we noticed.

File: cfgov/jinja2/v1/test-fixture/index.html [hunk 6]
```diff
@@ -125,7 +125,7 @@
     (b) In the <head> with the `defer` and `async` attributes
         set on the <script> tag.
 #}
-<script type="text/javascript" src="{{ static('js/atomic/' + atomic_type + suffix + '.js') }}"></script>
+<script type="text/javascript" src="{{ static('js/routes/on-demand/' + atomic_type + suffix + '.js') }}"></script>
```
Comment (Scotchester) [inline cfgov/jinja2/v1/test-fixture/index.html]: @jimmynotjim Do we need to revert this change?

File: frontend.sh [hunk 7]
```diff
@@ -85,11 +85,14 @@ clean_and_install() {
...
   echo 'Building project…'
-  gulp clean
   gulp build
 
   if [ "$cli_flag" = "production" ]; then
+    echo 'Running additional build steps for on-demand and Nemo assets.'
     gulp scripts:ondemand
+    gulp styles:ondemand
+    gulp scripts:nemo
+    gulp styles:nemo
```
Comment (Scotchester) [inline frontend.sh]: Added these extra gulp steps that are currently done in the Jenkins build job, but seems like they should be done here instead. Once this is merged, I'll PR their removal from the Jenkins job.

File: gulp/tasks/clean.js [hunk 8]
```diff
@@ -4,6 +4,18 @@ const configClean = require( '../config' ).clean;
...
+
+gulp.task( 'clean:css', () => {
+  // Clean CSS out of /cfgov/static_built/css/
+  del( configClean.css + '/**/*' );
+} );
+
+gulp.task( 'clean:js', () => {
+  // Clean JavaScript out of /cfgov/static_built/js/
+  del( configClean.js + '/**/*' );
+} );
```
Comment (Scotchester) [inline gulp/tasks/clean.js]: @sebworks I addressed your earlier comment by getting over my fear of anonymous arrow functions and having the format of this match the original `clean` task below. Let me know if that suits you.

File: package.json [hunk 9]
```diff
@@ -77,6 +77,7 @@
     "chai": "4.1.2",
     "chai-as-promised": "7.1.1",
     "cucumber": "3.0.6",
+    "gulp-changed": "3.1.0",
```
Comment (Scotchester) [inline package.json]: I accidentally duplicated this here in `devDependencies` by running `npm install --save-dev gulp-changed` when it's already in `dependencies`. Not sure how to fix with the lockfile system. Also, I know I've asked this before, so I'm sorry, but why are all our Gulp dependencies not `devDependencies`?

**Comment by Scotchester** [comment]
Alrighty! This is ready for a fresh review @sebworks, @anselmbradford. I left a few comments about some of the changes. Our changes in webpack config will definitely conflict with #3558. Not sure the best way to make that easy on ourselves.

**Comment by sebworks** [comment]
@Scotchester, how much time does this save?

File: cfgov/jinja2/v1/_layouts/base.html (hunk 5 above)
Comment (sebworks) [reply cfgov/jinja2/v1/_layouts/base.html]: yeah, it's still a typo.

File: config/webpack-config.js [hunk 10]
```diff
@@ -58,102 +53,76 @@ const COMMON_UGLIFY_CONFIG = new UglifyWebpackPlugin( {
...
-    new webpack.optimize.CommonsChunkPlugin( {
-      name: COMMON_BUNDLE_NAME
-    } ),
-    COMMON_UGLIFY_CONFIG ]
+    COMMON_UGLIFY_CONFIG
+  ]
 };
 
 module.exports = {
+  commonConf:            commonConf,
```
Comment (sebworks) [inline config/webpack-config.js]: you can just use shorthand for these:
`{ commonConf,
    externalConf,
   etc..
}`.

File: cfgov/jinja2/v1/_layouts/base.html (hunk 5 above)
Comment (Scotchester) [reply cfgov/jinja2/v1/_layouts/base.html]: Oof, wow :laughing:

File: package.json (hunk 9 above)
Comment (sebworks) [reply package.json]: The short answer is that they are used to create the build on Jenkins.

**Comment by sebworks** [comment]
💯, nice work on this.

File: package.json (hunk 9 above)
Comment (Scotchester) [reply package.json]: Jenkins runs tests, too, and all those things are in `devDependencies`.

**Comment by Scotchester** [comment]
@sebworks: As a couple reference points:

- A full run of `gulp styles` drops from about 10 seconds to under 1 second
- A full run of `gulp scripts` drops from about 23 seconds to about 4 seconds (not everything in the full scripts run can currently be put through gulp-changed, due to some file renaming that's happening, which gulp-changed can't handle)
- `gulp styles:modern` drops from roughly 2.3 seconds for a normal run to 34 ms if it's skipped
- `gulp scripts:modern` drops from 18 seconds to 133 ms

**Comment by sebworks** [comment]
@Scotchester, please check the sizes of the bundles. It looks like common.js has increased in size 5kb.

File: cfgov/jinja2/v1/test-fixture/index.html (hunk 6 above)
Comment (jimmynotjim) [reply cfgov/jinja2/v1/test-fixture/index.html]: I would think not. The real on-demand scripts do come from `/routes/on-demand`, only the old legacy pages required being built without common.js. Maybe @sebworks can confirm though.

**Comment by jimmynotjim** [comment]
@sebworks what are you comparing to? I'm getting the same exact file size between common built by this branch and common built by master, 23kb

**Comment by sebworks** [comment]
> @sebworks what are you comparing to? I'm getting the same exact file size between common built by this branch and common built by master, 23kb

I was getting 27kb, let me try again.

**Comment by sebworks** [comment]
@jimmynotjim, it looks good after running gulp clean.

**Comment by Scotchester** [comment]
Pushed a fix for `es5-shim.js` not getting named properly.

File: cfgov/jinja2/v1/test-fixture/index.html (hunk 6 above)
Comment (sebworks) [reply cfgov/jinja2/v1/test-fixture/index.html]: I don't think this file is used and can be deleted.
